Add Stages.md, revise Working Mode to include (#222)
whatwg/sg: Stages.md

> # Stages - a tool for encouraging attention to incubation features

Reviewer: Not sure if a policy needs a subtitle in this way; it would feel a bit out of place among the others on https://whatwg.org/policies I think. Maybe just `# Stages` would work.

Author: Fixed in https://github.com/whatwg/sg/pull/222/commits/dd195992df4f01dbd374454f6642065ec8a2bbe9.

> The WHATWG's "document reality" approach has been effective for nailing down HTML, DOM, XHR, and similar fundamental parts of the platform and improving interoperability and developer satisfaction. However, our process can sometimes be daunting for newcomers, who don't know how to reliably get implementer feedback or editor time commitment. This can discourage contributions, and it can lead to frustration when implementer/editor feedback arrives late in the process after the contributor feels "close to done".

Reviewer: ```suggestion
The WHATWG's "document reality" approach has been effective for nailing down [fundamental parts of the platform](https://spec.whatwg.org/) and improving interoperability and developer satisfaction. However, our process can sometimes be daunting for newcomers, who don't know how to reliably get implementer feedback or editor time commitment. This can discourage contributions, and it can lead to frustration when implementer/editor feedback arrives late in the process after the contributor feels "close to done".
```

The stage process described here is an additional tool designed to help provide a clear on-ramp for incubation efforts for contributors who aren't already editors, especially for additions to existing specs.  Stages are an optional, opt-in process that a feature proposer can use if they want to get more formal signals about how their feature is progressing. It asks for explicit implementer involvement at multiple stages, starting from notification that the problem is being worked on, then sign-off on the rough API and specification, and finally agreement on the full specification text. 

> Stage signals are also useful to the broader community, helping developers monitor the proposals that are moving through the various stages.  By explicitly signaling a proposal's progress, including implementer involvement, the community has a better idea of what is going on in WHATWG-land.  This tool is generally used for medium-to-large new features; it's not expected that explicit stage progression will be used for every feature.

Reviewer: ```suggestion
Stage signals are also useful to the broader community, helping developers monitor the proposals that are moving through the various stages.  By explicitly signaling a proposal's progress, including implementer involvement, the community has a better idea of what is going on in the WHATWG.

This tool is generally used for medium-to-large new features; it's not expected that explicit stage progression will be used for every feature.
```

# Details

> This document formalizes some signal checkpoints for new features in existing WHATWG specs. These checkpoints are modeled loosely on the [TC39 process](https://tc39.es/process-document/), which uses the concept of _stages_. Each subsequent stage implies a larger degree of consensus from the community, willingness to engage, and/or intent to implement and ship the feature in browser engines, and signals progress to web developers.  These checkpoints are not required for all features; this is simply a way to ensure that attention is appropriately drawn to incubations that have grown community support behind them.

Reviewer: ```suggestion
This document formalizes some signal checkpoints for new features in existing WHATWG Living Standards. These checkpoints are modeled loosely on the [TC39 process](https://tc39.es/process-document/), which uses the concept of _stages_. Each subsequent stage implies a larger degree of consensus from the community, willingness to engage and/or implement and ship the feature in browser engines, and signals progress to web developers.  These checkpoints are not required for all features; this is simply a way to ensure that attention is appropriately drawn to incubations that have grown community support behind them.
```

> Terminology:

Reviewer: ```suggestion
## Terminology
```

> * _Stage_: each successive stage is closer to a landed spec and interoperable implementation.  These stages are tracked on the issues in the relevant WHATWG Github repository, using labels ("stage 0", "stage 1", etc.)

Reviewer: ```suggestion
* _Stage_: each successive stage is closer to a landed spec and interoperable implementation.  These stages are tracked on the issues in the relevant WHATWG GitHub repository, using labels ("stage 0", "stage 1", etc.)
```

> * _Stage_: each successive stage is closer to a landed spec and interoperable implementation.  These stages are tracked on the issues in the relevant WHATWG Github repository, using labels ("stage 0", "stage 1", etc.)

Reviewer: Consider using `<dfn>` instead of `_`

Author: fixed in https://github.com/whatwg/sg/pull/222/commits/dd195992df4f01dbd374454f6642065ec8a2bbe9.

* _Browser engine_: an independent implementation of the web platform (e.g. Chromium, WebKit and Gecko at present).

> * _Implementation interest_: a direct signal from a browser engine that a feature is on the roadmap to implement, as defined in [working mode](https://whatwg.org/working-mode#additions)].

Reviewer: ```suggestion
* _Implementation interest_: a direct signal from a browser engine that a feature is on the roadmap to implement, as defined in [working mode](https://whatwg.org/working-mode#additions).
```

* _Browser engine representatives_: persons delegated by a browser engine to indicate implementation interest.
* _Proposal champion_: person/organization that proposed and will advance a feature through the stages, typically including writing the specification, researching use cases, and prototyping.

> Process:

Reviewer: ```suggestion
## Process
```

* Any proposal effectively starts at Stage 0 without any approvals, by a community member filing a new issue on a relevant WHATWG spec.
* In order to drive a proposal forward, it is expected that a community member will step up to champion the proposal, through answering questions asynchronously, providing feedback when requested, and consideration of important open questions at triage meetings.
* Stage level is tracked with labels on issues in the relevant specifications — “stage 0”, “stage 1”, etc.
* Stage labels (other than stage 0) should only be added by browser engine representatives, or by Editors on the relevant specification, after showing the necessary support for a level.
* **Advancing stages**

>     * The feature champion should bring the proposal to the WHATWG to advance to the next stage by setting an “Agenda+” label on the tracking issue, and showing up to (or ensuring someone will show up to) the next triage meeting to discuss. Advancing to a new stage requires support for the decision by at least two implementers (via their browser engine representatives), and there should not be any strong implementer objections ([per the working mode](https://whatwg.org/working-mode#additions)). This support can also be gathered in any public manner, e.g. GitHub issue comments, triage meetings, etc. 

Reviewer: The first sentence about agenda+ and triage meetings seems to contradict the last sentence about gathering support in any manner. Maybe it can be deleted?

Author: Although I agree the last sentence is not entirely necessary, it's not contradictory - the first sentence is about "how do you bring the proposal to the group", the last sentence is "how do you demonstrate support from implementers".

        * That the particular solution approach at the time of the signal or support will be the one eventually standardized (unless the advancement is to the Standardized stage)
    * A negative response to advancing a stage does **not** imply:
        * That the feature cannot continue in its current stage
        * That a future request for support to advance a stage with a revised solution won’t succeed
        * That all browsers will delay shipping the feature
        * That the feature cannot continue to be incubated outside of the WHATWG
    * A proposal can advance directly to later stages without going through the earlier stages, if it's already mature enough and has sufficient support to do so.
* **Regressing stages**
    * The browser engine representatives may review proposals and decide to regress stages—for example, if implementer support has been retracted for a particular solution (regressing from stage 3 to stage 2), or even if support has been retracted for solving the problem (i.e. a regression from stage 2 to stage 1). This situation is expected to be unusual, and should come with a very strong justification.

<table>
  <tr>

>    <td>

Reviewer: Make these th instead of td, and probably it'd be good to introduce thead and tbodys.

Author: Done in https://github.com/whatwg/sg/pull/222/commits/dd195992df4f01dbd374454f6642065ec8a2bbe9.

# Details

This document formalizes some signal checkpoints for new features in existing WHATWG specs. These checkpoints are modeled loosely on the [TC39 process](https://tc39.es/process-document/), which uses the concept of _stages_. Each subsequent stage implies a larger degree of consensus from the community, willingness to engage, and/or intent to implement and ship the feature in browser engines, and signals progress to web developers.  These checkpoints are not required for all features; this is simply a way to ensure that attention is appropriately drawn to incubations that have grown community support behind them.

Terminology:

* _Stage_: each successive stage is closer to a landed spec and interoperable implementation.  These stages are tracked on the issues in the relevant WHATWG Github repository, using labels ("stage 0", "stage 1", etc.)
* _Browser engine_: an independent implementation of the web platform (e.g. Chromium, WebKit and Gecko at present).

> * _Implementation interest_: a direct signal from a browser engine that a feature is on the roadmap to implement [Ed: same as in [current process](https://whatwg.org/working-mode#additions)].

Reviewer 2: Should that process still be called the "current" process once this one is also current?

Author: Good catch, restructured.

Reviewer 2: I don't see the change corresponding to this comment; have I missed it?

Author: Gaaah, no, sorry, my fault; I'd committed but not pushed a change that had this wrapped in it.  It's there now:

* _Implementation interest_: a direct signal from a browser engine that a feature is on the roadmap to implement, as defined in [working mode](https://whatwg.org/working-mode#additions)].

* _Browser engine representatives_: persons delegated by a browser engine to indicate implementation interest.
* _Proposal champion_: person/organization that proposed and will advance a feature through the stages, typically including writing the specification, researching use cases, and prototyping.

Process:

* Any proposal effectively starts at Stage 0 without any approvals, by a community member filing a new issue on a relevant WHATWG spec.
* In order to drive a proposal forward, it is expected that a community member will step up to champion the proposal, through answering questions asynchronously, providing feedback when requested, and consideration of important open questions at triage meetings.
* Stage level is tracked with labels on issues in the relevant specifications — “stage 0”, “stage 1”, etc.
* Stage labels (other than stage 0) should only be added by browser engine representatives, or by Editors on the relevant specification, after showing the necessary support for a level.
* **Advancing stages**

>     * The feature champion should bring the proposal to the WHATWG to advance to the next stage by setting an “Agenda+” label on the tracking issue, and showing up to (or ensuring someone will show up to) the next triage meeting to discuss. Advancing to a new stagerequires support for the decision by at least two implementers (via their browser engine representatives), and there should not be any strong implementer objections ([per the working mode](https://whatwg.org/working-mode#additions)). This support can also be gathered in any public manner, e.g. GitHub issue comments, triage meetings, etc. 

Reviewer 2: ```suggestion
    * The feature champion should bring the proposal to the WHATWG to advance to the next stage by setting an “Agenda+” label on the tracking issue, and showing up to (or ensuring someone will show up to) the next triage meeting to discuss. Advancing to a new stage requires support for the decision by at least two implementers (via their browser engine representatives), and there should not be any strong implementer objections ([per the working mode](https://whatwg.org/working-mode#additions)). This support can also be gathered in any public manner, e.g. GitHub issue comments, triage meetings, etc. 
```

<table>
  <tr>
   <td>
Stage Name
   </td>
   <td>Entrance criteria
   </td>
   <td>This stage signifies
   </td>
   <td>Spec quality at entrance
   </td>
  </tr>
  <tr>
   <td>Stage 0
<p>
(Proposal)
   </td>
   <td>(None)
   </td>
   <td>That the proposer intends to use the stage process to work on their idea, with the goal of it landing in WHATWG Living Standard(s).
   </td>
   <td>An explainer describing the  user/developer problem to be solved, including sketching use cases and scenarios. This explainer can exist anywhere, including e.g. a GitHub issue, or a personal repository.
   </td>
  </tr>
  <tr>
   <td>Stage 1
<p>
(Incubation)
   </td>
   <td>A comprehensive explainer for the feature, in a standards organization-approved incubation venue such as a <a href="https://www.w3.org/community/groups/">W3C CG</a> or a branch of an existing WHATWG Standard.
<p>
Consensus that the WHATWG is interested in exploring solutions in this problem space.
<p>
(At least) one implementer interested in doing prototyping work.
<p>

> Identification of the feature's champion. \

Reviewer 2: These stray backslashes confused me. Are they intended to be there? It looks like they were intended to escape newlines in some different document format.

Author: Yeesh, yes, the bullets went haywire.  Fixed.
